### DISCUSSION TITLE
Scala/Slick AsyncExecutor instrumentation

### DISCUSSION THREAD

**Discussion by dboreham** [open-telemetry/opentelemetry-java-instrumentation#5299] [Q&A]
Hello, does the agent as-is automatically instrument applications using Slick? It has been reported to me that it doesn't (parent span is lost after the context switch to a thread in Slick's AsyncExecutor), however I see some Slick-related code, and a test in the agent: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/scala-fork-join-2.8/javaagent/src/slickTest/scala/SlickUtils.scala which suggests that perhaps it is expected to work. I wondered if perhaps the agent's Scala support was developed in the past when Slick used the stock Scala ForkJoinPool and things had rotted in the meantime.

Anyway, it'd be great to hear any thoughts, pointers, etc on this before I dive in. Thanks.

**Comment by trask** [comment]
hey @dboreham! the `scala-fork-join-2.8` module is just targeted at the legacy Scala ForkJoinPool, which was deprecated in 2.12 IIUC

if Slick has its own ExecutorService implementation, you may be able to just add that to https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/21f968abb45190e280e0cde3ea1468b36d4524f8/instrumentation/executors/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/javaconcurrent/AbstractExecutorInstrumentation.java#L57-L101

in this case, adding a test with `testLibrary` Slick will make sure CI always runs against both that base version of slick and the "latest" version of Slick, which will ensure it doesn't rot again, e.g.

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/21f968abb45190e280e0cde3ea1468b36d4524f8/instrumentation/cdi-testing/build.gradle.kts#L6-L8

if not, you may need to implement something similar to the ExecutorService instrumentation but for Slick's specific classes

in this case, check out https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation.md (and https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md)

**Reply by dboreham** [reply]
Ok thanks. I wondered about adding the executor to that list. But Slick actually just uses the stock `java.util.concurrent.ThreadPoolExecutor` now, which prompts the question: why doesn't it work already? I will investigate and report back...
